rfcs/react-components/convergence/component-transitions-on-mount-or-unmount.md

```diff
@@ -0,0 +1,239 @@
+
+  /*
+   * Determines whether the component should be displayed on the screen.
+   *
+   * This property will evaluate to `true` when the provided presence value is true.
+   * Once the provided value changes to false, the hook will monitor the transition/animation
+   * completion and subsequently set this property to false.
+   */
+  shouldRender,
+
+  /*
+   * Indicates whether the component is currently rendered and visible.
+   *
+   * This flag will be set to `true` one frame after the shouldRender value becomes true.
+   * It will be set to `false` when the specified presence value changes to false.
+   */
+  visible,
+
+  /*
+   * Indicates whether the component is currently entering the screen.
+   *
+   * This will be `true` when the element is entering the screen. This happens
+   * one frame after `visible` value becomes true.
+   * It becomes `false` when the transition/animation ends.
+   */
+  entering,
+
+  /*
+   * Whether the component is leaving the screen
+   *
+   * This property is set to `true` one frame after the `visible` prop becomes true,
+   * indicating that the element is in the process of entering the screen.
+   * It becomes `false` when the transition/animation ends.
+   * It is set to false once the transition or animation for the component has completed.
+   */
+  exiting,
+
+  /*
+   * Indicates whether the component is currently undergoing animation.
+   * This property will be set to `true` when the component is either entering or exiting.
+   */
+  animating,
+} = useMotionPresence(open);
+```
+
+Events can be provided as a second argument of the hook.:
+
+```js
+  const presence = useMotionPresence(open, {
+    onEntered: () => {
+      // Called when the element finished the "enter" animation/transition
+    };
+    onExited: () => {
+      // Called when the element finished the "leave" animation/transition
+    };
+  })
+```
+
+Other events can be implemented, like `onBeforeEnter` or `onBeforeLeave`.
+
+#### Background research
+
+- Vue.js [Transition component](https://vuejs.org/guide/built-ins/transition.html#javascript-hooks). The [implementation](https://github.com/vuejs/core/blob/main/packages/runtime-dom/src/components/Transition.ts) CSS transitions and animations and provides classes to the components based on the detected transitions.
+- Radix UI has an internal package called [Presence](https://www.npmjs.com/package/@radix-ui/react-presence), which addresses the same issue. To handle this, they offer a [`usePresence` hook and a Presence component](https://github.com/radix-ui/primitives/blob/main/packages/react/presence/src/Presence.tsx). The implementation relies on animations only so they only need to determine whether the element should be present or not on the screen.
+
+#### Pros
+
+- 👍 Hook based solution.
+- 👍 Griffel styles can be used normally to create motion.
+- 👍 Easy styling of components based on their state.
+- 👍 Offers the flexibility to declare both transitions and animations.
```

Comment:
It would be nice to see an example with CSS animations.

Reply:
Added to the examples.